jetbrains-core/src/software/aws/toolkits/jetbrains/services/telemetry/ClientTelemetryService.kt

```diff
@@ -27,23 +28,55 @@ import java.net.URI
 import java.time.Duration
 import java.time.Instant
 
-interface ClientTelemetryService : Disposable
+interface ClientTelemetryService : Disposable {
+    fun recordMetric(metricNamespace: String): Metric
+}
 
-class DefaultClientTelemetryService(sdkClient: AwsSdkClient, regionProvider: ToolkitRegionProvider) : ClientTelemetryService {
+sealed class DefaultClientTelemetryService(
+    private val sdkClient: AwsSdkClient,
+    private val regionProvider: ToolkitRegionProvider
+) : ClientTelemetryService {
     private lateinit var startupTime: Instant
-    private val publisher: MetricsPublisher
+    private var publisher: MetricsPublisher = NoOpMetricsPublisher()
 
     init {
         Disposer.register(ApplicationManager.getApplication(), sdkClient)
 
-        val clientManager = ServiceManager.getService(
+        recordMetric("ToolkitStart").use {
+            startupTime = it.createTime
+            // Metadata must be attached to a metric entry. There are no metric data points that we need to log for
+            // this event, so we log a placeholder entry with the relevant metadata.
+            it.addMetricEntry("metadata") {
+                value(0.0)
+                unit(MetricUnit.COUNT)
+                // TODO
+                // val isFirstRun: boolean = ...
+                // metadata("isFirstRun", isFirstRun.toString())
+            }
+        }
+    }
+
+    @Synchronized
+    override fun recordMetric(metricNamespace: String): Metric {
+        val settings = AwsSettings.getInstance()
+        // if telemetry changed from disabled to enabled
+        if (settings.isTelemetryEnabled && publisher is NoOpMetricsPublisher) {
+            try {
+                publisher.shutdown()
+            } finally {
+                publisher = NoOpMetricsPublisher()
```

Comment:
Is this logic correct here? When the `telemetry` is enabled, we should set the `publisher` to a real publisher, ie `BatchingMetricsPublisher`, right ?

Reply:
Good catch, the `if` blocks should be swapped.